File: 2870-minimum-number-of-operations-to-make-array-empty/min_operations.js
function minOperations(nums) {
    let res = 0;
    const counts = getCharacterFrequencyMap(nums);

    for(let [char, count] of counts) {
        if(count == 1) return -1;
        res += Math.floor(count / 3) + (count % 3 === 0 ? 0 : 1);
    }

    return res;
}

function getCharacterFrequencyMap(word) {
    const freqMap = new Map();
    for (const char of word) {
        freqMap.set(char, (freqMap.get(char) || 0) + 1);
    }
    return freqMap;
}

// Example usage:
const nums1 = [2, 3, 3, 2, 2, 4, 2, 3, 4];
const result1 = minOperations(nums1);
console.log(result1); // Output: 4

const nums2 = [2, 1, 2, 2, 3, 3];
const result2 = minOperations(nums2);
console.log(result2); // Output: -1

File: 2870-minimum-number-of-operations-to-make-array-empty/min_operations.py
from collections import Counter
from typing import List

class Solution:
    def minOperations(self, nums: List[int]) -> int:
        res = 0
        counts = Counter(nums)
        for count in counts.values():
            if count == 1:
                return -1
            res += count // 3 + (1 if count % 3 else 0)
        return res 

# Example usage:
solution = Solution()    
nums1 = [2, 3, 3, 2, 2, 4, 2, 3, 4]
result1 = solution.minOperations(nums1)
print(result1) # Output: 4

nums2 = [2, 1, 2, 2, 3, 3]
result2 = solution.minOperations(nums2)
print(result2) # Output: -1